File: mnheme_ports/go/mnheme/memorydb.go
// mnheme/go/mnheme/memorydb.go
// =============================
// MemoryDB — public API
//
// Usage:
//   db, _ := mnheme.Open("mente.mnheme")
//   mem, _ := db.Remember("Debito", "ansia", "Ho firmato il mutuo.", "", nil)
//   memories, _ := db.Recall("Debito", "", 0, false)

package mnheme

import (
	"crypto/sha256"
	"encoding/json"
	"fmt"
	"strings"
	"time"

	"github.com/google/uuid"
)

// Valid feelings
var validFeelings = map[string]struct{}{
	"gioia": {}, "tristezza": {}, "rabbia": {}, "paura": {},
	"nostalgia": {}, "amore": {}, "malinconia": {}, "serenità": {},
	"sorpresa": {}, "ansia": {}, "gratitudine": {}, "vergogna": {},
	"orgoglio": {}, "noia": {}, "curiosità": {},
}

// Memory is an immutable snapshot of a stored memory record.
type Memory struct {
	MemoryID  string   `json:"memory_id"`
	Concept   string   `json:"concept"`
	Feeling   string   `json:"feeling"`
	MediaType string   `json:"media_type"`
	Content   string   `json:"content"`
	Note      string   `json:"note"`
	Tags      []string `json:"tags"`
	Timestamp string   `json:"timestamp"`
	Checksum  string   `json:"checksum"`
}

// MemoryDB is the main database object.
type MemoryDB struct {
	storage *StorageEngine
	index   *IndexEngine
	path    string
}

// Open opens or creates the database at path.
// Cold start: scans the log and rebuilds indexes.
func Open(path string) (*MemoryDB, error) {
	storage, err := NewStorageEngine(path)
	if err != nil {
		return nil, fmt.Errorf("mnheme.Open: %w", err)
	}
	index   := NewIndexEngine()
	records, err := storage.Scan()
	if err != nil {
		return nil, fmt.Errorf("mnheme.Open scan: %w", err)
	}
	index.Rebuild(records)
	return &MemoryDB{storage: storage, index: index, path: path}, nil
}

// ── Write ─────────────────────────────────────────────────────────

// Remember appends a new memory to the database.
func (db *MemoryDB) Remember(
	concept   string,
	feeling   string,
	content   string,
	note      string,
	tags      []string,
	mediaType string,
) (*Memory, error) {
	concept = strings.TrimSpace(concept)
	if concept == "" {
		return nil, fmt.Errorf("concept cannot be empty")
	}
	if content == "" {
		return nil, fmt.Errorf("content cannot be empty")
	}
	if mediaType == "" {
		mediaType = "text"
	}
	if _, ok := validFeelings[feeling]; !ok {
		return nil, fmt.Errorf("invalid feeling: %q", feeling)
	}
	if tags == nil {
		tags = []string{}
	}

	memID     := uuid.NewString()
	timestamp := time.Now().UTC().Format(time.RFC3339Nano)
	checksum  := fmt.Sprintf("%x", sha256.Sum256([]byte(content)))

	record := map[string]any{
		"memory_id":  memID,
		"concept":    concept,
		"feeling":    feeling,
		"media_type": mediaType,
		"content":    content,
		"note":       note,
		"tags":       tags,
		"timestamp":  timestamp,
		"checksum":   checksum,
	}

	offset, err := db.storage.Append(record)
	if err != nil {
		return nil, fmt.Errorf("remember: %w", err)
	}
	db.index.IndexRecord(offset, record)

	return &Memory{
		MemoryID:  memID,
		Concept:   concept,
		Feeling:   feeling,
		MediaType: mediaType,
		Content:   content,
		Note:      note,
		Tags:      tags,
		Timestamp: timestamp,
		Checksum:  checksum,
	}, nil
}

// ── Read ──────────────────────────────────────────────────────────

// Recall returns memories for a concept, optionally filtered by feeling.
// limit=0 means no limit. Pass oldestFirst=true for ascending order.
func (db *MemoryDB) Recall(concept, feeling string, limit int, oldestFirst bool) ([]*Memory, error) {
	offsets := db.index.OffsetsByConcept(concept, feeling, oldestFirst)
	return db.loadOffsets(offsets, limit)
}

// RecallByFeeling returns all memories with the given feeling.
func (db *MemoryDB) RecallByFeeling(feeling string, limit int, oldestFirst bool) ([]*Memory, error) {
	offsets := db.index.OffsetsByFeeling(feeling, oldestFirst)
	return db.loadOffsets(offsets, limit)
}

// RecallAll returns every memory in the database.
func (db *MemoryDB) RecallAll(limit int, oldestFirst bool) ([]*Memory, error) {
	offsets := db.index.AllOffsets(oldestFirst)
	return db.loadOffsets(offsets, limit)
}

// Search performs a full-text scan across content, concept, and note.
func (db *MemoryDB) Search(
	text string,
	inContent, inConcept, inNote bool,
	limit int,
) ([]*Memory, error) {
	needle  := strings.ToLower(text)
	records, err := db.storage.Scan()
	if err != nil {
		return nil, err
	}
	var results []*Memory
	// Iterate in reverse (most recent first)
	for i := len(records) - 1; i >= 0; i-- {
		r := records[i].Record
		matched := false
		if inContent && strings.Contains(strings.ToLower(strVal(r, "content")), needle) {
			matched = true
		}
		if inConcept && strings.Contains(strings.ToLower(strVal(r, "concept")), needle) {
			matched = true
		}
		if inNote && strings.Contains(strings.ToLower(strVal(r, "note")), needle) {
			matched = true
		}
		if matched {
			results = append(results, recordToMemory(r))
			if limit > 0 && len(results) >= limit {
				break
			}
		}
	}
	return results, nil
}

// ── Stats ─────────────────────────────────────────────────────────

// Count returns number of memories, optionally filtered.
// Pass empty strings to skip filters.
func (db *MemoryDB) Count(concept, feeling string) int {
	return db.index.Count(concept, feeling)
}

// FeelingDistribution returns feeling → count map.
func (db *MemoryDB) FeelingDistribution() map[string]int {
	return db.index.FeelingDistribution()
}

// ListConcepts returns all known concepts sorted alphabetically.
func (db *MemoryDB) ListConcepts() []string {
	return db.index.AllConcepts()
}

// ConceptTimeline returns the emotional evolution of a concept over time.
func (db *MemoryDB) ConceptTimeline(concept string) ([]map[string]any, error) {
	offsets := db.index.TimelineOffsets(concept)
	var result []map[string]any
	for _, off := range offsets {
		r, err := db.storage.ReadAt(off)
		if err != nil {
			continue
		}
		result = append(result, map[string]any{
			"timestamp": strVal(r, "timestamp"),
			"feeling":   strVal(r, "feeling"),
			"note":      strVal(r, "note"),
			"tags":      r["tags"],
		})
	}
	return result, nil
}

// ── Export ────────────────────────────────────────────────────────

// ExportJSON serialises all memories to a JSON string.
func (db *MemoryDB) ExportJSON() (string, error) {
	memories, err := db.RecallAll(0, true)
	if err != nil {
		return "", err
	}
	out := map[string]any{
		"exported_at": time.Now().UTC().Format(time.RFC3339),
		"memories":    memories,
	}
	b, err := json.MarshalIndent(out, "", "  ")
	return string(b), err
}

// ── Internal ──────────────────────────────────────────────────────

func (db *MemoryDB) loadOffsets(offsets []uint64, limit int) ([]*Memory, error) {
	if limit > 0 && limit < len(offsets) {
		offsets = offsets[:limit]
	}
	result := make([]*Memory, 0, len(offsets))
	for _, off := range offsets {
		r, err := db.storage.ReadAt(off)
		if err != nil {
			continue
		}
		result = append(result, recordToMemory(r))
	}
	return result, nil
}

func recordToMemory(r map[string]any) *Memory {
	m := &Memory{
		MemoryID:  strVal(r, "memory_id"),
		Concept:   strVal(r, "concept"),
		Feeling:   strVal(r, "feeling"),
		MediaType: strVal(r, "media_type"),
		Content:   strVal(r, "content"),
		Note:      strVal(r, "note"),
		Timestamp: strVal(r, "timestamp"),
		Checksum:  strVal(r, "checksum"),
	}
	if tags, ok := r["tags"].([]any); ok {
		for _, t := range tags {
			if s, ok := t.(string); ok {
				m.Tags = append(m.Tags, s)
			}
		}
	}
	return m
}

// go.mod (module declaration)
// module github.com/aatel-license/mnheme
// go 1.22
// require github.com/google/uuid v1.6.0

File: mnheme_ports/java/mnheme/MemoryDB.java
// mnheme/java/mnheme/MemoryDB.java
// ==================================
// MemoryDB — public API for MNHEME
//
// Usage:
//   var db = new MemoryDB(Path.of("mente.mnheme"));
//   var mem = db.remember("Debito", Feeling.ANSIA, "Ho firmato il mutuo.");
//   var list = db.recall("Debito");
//
// Build (Maven):
//   <dependency>
//     <groupId>com.fasterxml.jackson.core</groupId>
//     <artifactId>jackson-databind</artifactId>
//     <version>2.17.0</version>
//   </dependency>
//   <dependency>
//     <groupId>com.github.f4b6a3</groupId>
//     <artifactId>uuid-creator</artifactId>
//     <version>5.3.3</version>
//   </dependency>

package mnheme;

import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.IOException;
import java.nio.file.Path;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.time.Instant;
import java.util.*;
import java.util.stream.Collectors;

public class MemoryDB implements AutoCloseable {

    private final StorageEngine storage;
    private final IndexEngine   index;
    private final String        dbPath;
    private final ObjectMapper  mapper = new ObjectMapper();

    /**
     * Open or create the database at path.
     * Cold start: scans the log and rebuilds all indexes.
     */
    public MemoryDB(Path path) throws IOException {
        this.dbPath  = path.toString();
        this.storage = new StorageEngine(path);
        this.index   = new IndexEngine();
        var records  = storage.scan();
        index.rebuild(records);
    }

    // ── Write ──────────────────────────────────────────────────

    /**
     * Append a new memory. Throws if concept/content is empty or feeling invalid.
     */
    public Memory remember(
            String       concept,
            Feeling      feeling,
            String       content,
            String       note,
            List<String> tags,
            String       mediaType
    ) throws IOException {
        concept   = concept.strip();
        if (concept.isEmpty()) throw new IllegalArgumentException("Concept cannot be empty");
        if (content.isEmpty()) throw new IllegalArgumentException("Content cannot be empty");
        if (note      == null) note      = "";
        if (tags      == null) tags      = List.of();
        if (mediaType == null) mediaType = "text";

        String memoryId  = UUID.randomUUID().toString();
        String timestamp = Instant.now().toString();
        String checksum  = sha256(content);

        Map<String, Object> record = new LinkedHashMap<>();
        record.put("memory_id",  memoryId);
        record.put("concept",    concept);
        record.put("feeling",    feeling.getValue());
        record.put("media_type", mediaType);
        record.put("content",    content);
        record.put("note",       note);
        record.put("tags",       tags);
        record.put("timestamp",  timestamp);
        record.put("checksum",   checksum);

        long offset = storage.append(record);
        index.indexRecord(offset, record);

        return new Memory(memoryId, concept, feeling.getValue(), mediaType,
                          content, note, List.copyOf(tags), timestamp, checksum);
    }

    /** Convenience: remember with defaults (no note, no tags). */
    public Memory remember(String concept, Feeling feeling, String content) throws IOException {
        return remember(concept, feeling, content, "", null, "text");
    }

    // ── Read ───────────────────────────────────────────────────

    /** Recall memories by concept. Most recent first. */
    public List<Memory> recall(String concept) throws IOException {
        return recall(concept, null, -1, false);
    }

    public List<Memory> recall(
            String  concept,
            Feeling feeling,
            int     limit,
            boolean oldestFirst
    ) throws IOException {
        String fVal = feeling != null ? feeling.getValue() : null;
        var offsets = index.offsetsByConcept(concept.strip(), fVal, oldestFirst);
        return loadOffsets(offsets, limit);
    }

    public List<Memory> recallByFeeling(Feeling feeling) throws IOException {
        return recallByFeeling(feeling, -1, false);
    }

    public List<Memory> recallByFeeling(
            Feeling feeling, int limit, boolean oldestFirst
    ) throws IOException {
        var offsets = index.offsetsByFeeling(feeling.getValue(), oldestFirst);
        return loadOffsets(offsets, limit);
    }

    public List<Memory> recallAll() throws IOException {
        return recallAll(-1, false);
    }

    public List<Memory> recallAll(int limit, boolean oldestFirst) throws IOException {
        var offsets = index.allOffsets(oldestFirst);
        return loadOffsets(offsets, limit);
    }

    /**
     * Full-text search across content, concept, note.
     */
    public List<Memory> search(String text) throws IOException {
        return search(text, true, true, true, -1);
    }

    public List<Memory> search(
            String  text,
            boolean inContent, boolean inConcept, boolean inNote,
            int     limit
    ) throws IOException {
        String needle  = text.toLowerCase(Locale.ROOT);
        var    records = storage.scan();
        Collections.reverse(records); // most recent first
        List<Memory> results = new ArrayList<>();

        for (var sr : records) {
            var r = sr.record();
            boolean matched = false;
            if (inContent && strVal(r,"content").toLowerCase().contains(needle)) matched = true;
            if (inConcept && strVal(r,"concept").toLowerCase().contains(needle)) matched = true;
            if (inNote    && strVal(r,"note"   ).toLowerCase().contains(needle)) matched = true;
            if (matched) {
                results.add(Memory.fromMap(r));
                if (limit > 0 && results.size() >= limit) break;
            }
        }
        return results;
    }

    // ── Stats ──────────────────────────────────────────────────

    public int count() { return index.count(null, null); }

    public int count(String concept, Feeling feeling) {
        return index.count(concept, feeling != null ? feeling.getValue() : null);
    }

    public Map<String, Integer> feelingDistribution() {
        return index.feelingDistribution();
    }

    public List<String> listConcepts() {
        return index.allConcepts();
    }

    public List<Map<String, Object>> conceptTimeline(String concept) throws IOException {
        var offsets = index.timelineOffsets(concept.strip());
        List<Map<String, Object>> result = new ArrayList<>();
        for (long off : offsets) {
            var r = storage.readAt(off);
            if (r == null) continue;
            result.add(Map.of(
                "timestamp", strVal(r, "timestamp"),
                "feeling",   strVal(r, "feeling"),
                "note",      strVal(r, "note"),
                "tags",      r.getOrDefault("tags", List.of())
            ));
        }
        return result;
    }

    // ── Export ─────────────────────────────────────────────────

    public String exportJson() throws IOException {
        var memories = recallAll(- 1, true).stream()
                .map(Memory::toMap)
                .collect(Collectors.toList());
        var out = Map.of(
            "exported_at", Instant.now().toString(),
            "memories",    memories
        );
        return mapper.writerWithDefaultPrettyPrinter().writeValueAsString(out);
    }

    // ── Internal ───────────────────────────────────────────────

    private List<Memory> loadOffsets(List<Long> offsets, int limit) throws IOException {
        int n = (limit > 0) ? Math.min(limit, offsets.size()) : offsets.size();
        List<Memory> result = new ArrayList<>(n);
        for (int i = 0; i < n; i++) {
            var r = storage.readAt(offsets.get(i));
            if (r != null) result.add(Memory.fromMap(r));
        }
        return result;
    }

    private static String strVal(Map<String, Object> m, String key) {
        Object v = m.get(key);
        return v instanceof String s ? s : "";
    }

    private static String sha256(String input) {
        try {
            var digest = MessageDigest.getInstance("SHA-256");
            byte[] hash = digest.digest(input.getBytes(java.nio.charset.StandardCharsets.UTF_8));
            var sb = new StringBuilder();
            for (byte b : hash) sb.append(String.format("%02x", b));
            return sb.toString();
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public void close() { /* no persistent connection to close */ }
}
